airbyte-bootloader/src/main/java/io/airbyte/bootloader/ProtocolVersionChecker.java

```diff
@@ -0,0 +1,218 @@
+  }
+
+  protected Set<UUID> projectRemainingConflictsAfterConnectorUpgrades(final AirbyteProtocolVersionRange targetRange,
+                                                                      final Set<UUID> initialConflicts,
+                                                                      final ActorType actorType) {
+    if (initialConflicts.isEmpty()) {
+      return Set.of();
+    }
+
+    final Set<UUID> upgradedSourceDefs = getProtocolVersionsForActorDefinitions(actorType)
+        // Keep definition ids if the protocol version will fall into the new supported range
+        .filter(e -> initialConflicts.contains(e.getKey()) && targetRange.isSupported(e.getValue()))
+        .map(Entry::getKey)
+        .collect(Collectors.toSet());
+
+    // Get the set of source definitions that will still have conflict after the connector upgrades
+    final Set<UUID> remainingConflicts = new HashSet<>(initialConflicts);
+    remainingConflicts.removeAll(upgradedSourceDefs);
+    return remainingConflicts;
+  }
+
+  protected Stream<Entry<UUID, Version>> getProtocolVersionsForActorDefinitions(final ActorType actorType) {
+    Stream<Entry<UUID, Version>> stream;
+    if (actorType == ActorType.SOURCE) {
+      stream = definitionsProvider.getSourceDefinitions()
+          .stream()
+          .map(def -> Map.entry(def.getSourceDefinitionId(), AirbyteProtocolVersion.getWithDefault(def.getSpec().getProtocolVersion())));
+    } else {
+      stream = definitionsProvider.getDestinationDefinitions()
+          .stream()
+          .map(def -> Map.entry(def.getDestinationDefinitionId(), AirbyteProtocolVersion.getWithDefault(def.getSpec().getProtocolVersion())));
+    }
+    return stream;
+  }
+
+  private Stream<String> formatActorDefinitionForLogging(final Set<UUID> remainingDestConflicts, final Set<UUID> remainingSourceConflicts) {
+    return Stream.concat(
+        remainingSourceConflicts.stream().map(defId -> {
+          final StandardSourceDefinition sourceDef;
+          try {
+            sourceDef = configRepository.getStandardSourceDefinition(defId);
+            return String.format("Source: %s: %s: protocol version: %s",
+                sourceDef.getSourceDefinitionId(), sourceDef.getName(), sourceDef.getProtocolVersion());
+          } catch (Exception e) {
+            log.info("Failed to getStandardSourceDefinition for " + defId, e);
+            return String.format("Source: %s: Failed to fetch details...", defId);
+          }
+        }),
+        remainingDestConflicts.stream().map(defId -> {
+          try {
+            final StandardDestinationDefinition destDef = configRepository.getStandardDestinationDefinition(defId);
+            return String.format("Destination: %s: %s: protocol version: %s",
+                destDef.getDestinationDefinitionId(), destDef.getName(), destDef.getProtocolVersion());
+          } catch (Exception e) {
+            log.info("Failed to getStandardDestinationDefinition for " + defId, e);
```

Comment:
Nit:  use placeholders instead of string concatenation in log message.

Reply:
done